### ISSUE TITLE
Option `WithAutoReconnectDelay` fails to reconnect when unauthorized.

### ISSUE THREAD

**Issue by jechtom** [dotnet/MQTTnet#1810]
Authorization failure breaks the managed client and it will never try to reconnect.

This happens if I use `ManagedMqttClient` with the feature `AutoReconnectDelay` ([tuned on by default](https://github.com/dotnet/MQTTnet/blob/9295626503231e50310aba104ed94e216ba4e95a/Source/MQTTnet.Extensions.ManagedClient/ManagedMqttClientOptions.cs#L15)) together with client option `WithoutThrowOnNonSuccessfulConnectResponse` (expected to be [turned on by default in the future](https://github.com/dotnet/MQTTnet/blob/9295626503231e50310aba104ed94e216ba4e95a/Source/MQTTnet/Client/Options/MqttClientOptionsBuilder.cs#L84)).

### Which component is your bug related to?

- Client  (4.2.1.781, latest on 2023-08-06)
- ManagedClient (4.2.1.781, latest on 2023-08-06)

## To Reproduce

1. Have any MQTT server that requires authentication.
2. Use (currently latest) `MQTTnet.Extensions.ManagedClient 4.2.1.781`.
3. Run this code:
```csharp
            var mqttClientOptions = new ManagedMqttClientOptionsBuilder()
                //.WithAutoReconnectDelay(TimeSpan.FromSeconds(5)) <- no needed, turned on by default
                .WithClientOptions(b => b
                    .WithCredentials("aaaaaaaaaa invalid credentials")
                    .WithoutThrowOnNonSuccessfulConnectResponse() // will be enabled by default in the future versions
                    .WithConnectionUri("wss://address-of-you-mqtt-server")
                ).Build();

            var mqttClient = new MqttFactory().CreateManagedMqttClient();
            await mqttClient.StartAsync(mqttClientOptions).ConfigureAwait(false);
            await Task.Delay(Timeout.Infinite);
```
5. Run the code. 
6. Client will contact the server only once and will not retry.

### Expected behavior
The client should try to reconnect in regular intervals (5s by default).
Instead if will internally fail on error: 
```
Not allowed to connect while connect/disconnect is pending.
```

## Full code example with logging

Full code:
```csharp
using MQTTnet.Extensions.ManagedClient;
using MQTTnet;
using System.Diagnostics;

namespace MqttReconnectTest
{
    internal class Program
    {
        static async Task Main(string[] args)
        {
            Stopwatch sw = Stopwatch.StartNew();

            Task Print(string text)
            {
                Console.WriteLine($"[{sw.Elapsed:s\\.fff}s] {text}");
                return Task.CompletedTask;
            }

            await Print("Starting.");

            var mqttClientOptions = new ManagedMqttClientOptionsBuilder()
                //.WithAutoReconnectDelay(TimeSpan.FromSeconds(5)) <- no e, turned on by default
                .WithClientOptions(b => b
                    .WithCredentials("aaaaaaaaaa invalid credentials")
                    .WithoutThrowOnNonSuccessfulConnectResponse() // will be enabled by default in the future versions
                    .WithConnectionUri("wss://localhost:6001/servicebus")
                ).Build();

            var mqttClient = new MqttFactory().CreateManagedMqttClient();

         
            mqttClient.ConnectedAsync += (t) => Print("ConnectedAsync.");
            mqttClient.ConnectingFailedAsync += (t) => Print($"ConnectingFailedAsync: {t.Exception}");
            mqttClient.ConnectionStateChangedAsync += (t) => Print("ConnectionStateChangedAsync.");
            mqttClient.DisconnectedAsync += (t) => Print("DisconnectedAsync.");

            await mqttClient.StartAsync(mqttClientOptions).ConfigureAwait(false);

            await Task.Delay(Timeout.Infinite);
        }
    }
}
```
Output:
```
[0.000s] Starting.
[0.209s] ConnectingFailedAsync: MQTTnet.Exceptions.MqttCommunicationException: Client connected but server denied connection with reason 'NotAuthorized'.
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
[5.217s] ConnectingFailedAsync: System.InvalidOperationException: Not allowed to connect while connect/disconnect is pending.
   at MQTTnet.Client.MqttClient.ConnectAsync(MqttClientOptions options, CancellationToken cancellationToken)
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
[10.230s] ConnectingFailedAsync: System.InvalidOperationException: Not allowed to connect while connect/disconnect is pending.
   at MQTTnet.Client.MqttClient.ConnectAsync(MqttClientOptions options, CancellationToken cancellationToken)
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
[15.241s] ConnectingFailedAsync: System.InvalidOperationException: Not allowed to connect while connect/disconnect is pending.
   at MQTTnet.Client.MqttClient.ConnectAsync(MqttClientOptions options, CancellationToken cancellationToken)
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
[20.243s] ConnectingFailedAsync: System.InvalidOperationException: Not allowed to connect while connect/disconnect is pending.
   at MQTTnet.Client.MqttClient.ConnectAsync(MqttClientOptions options, CancellationToken cancellationToken)
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
..........
```

**Comment by jechtom** [comment]
Seems like bug in `MqttClient` in the method `ConnectAsync`. I tried to fix it in #1810.

**Comment by jechtom** [comment]
Fixed in (merge) https://github.com/dotnet/MQTTnet/releases/tag/v4.3.0.858

**Comment by jwbrothers** [comment]
@jechtom I am still seeing this issue with 4.3.3.952.  The odd thing is I do not see this log message at all with Warning level tracing enabled:

` _logger.Warning("Connecting failed: {0}", connectResult.ResultCode);`

Will need to debug through this more tomorrow, but when debugging last week prior to upgrading to the latest version of the library the connection status was 'Connecting'.

**Comment by jechtom** [comment]
@jwbrothers Tested again with example code using .NET 8.

Version `4.2.1.781` before the fix:
```
[0.000s] Starting.
[3.869s] ConnectingFailedAsync: MQTTnet.Exceptions.MqttCommunicationException: Client connected but server denied connection with reason 'NotAuthorized'.
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
[8.874s] ConnectingFailedAsync: System.InvalidOperationException: Not allowed to connect while connect/disconnect is pending.
   at MQTTnet.Client.MqttClient.ConnectAsync(MqttClientOptions options, CancellationToken cancellationToken)
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
[13.883s] ConnectingFailedAsync: System.InvalidOperationException: Not allowed to connect while connect/disconnect is pending.
   at MQTTnet.Client.MqttClient.ConnectAsync(MqttClientOptions options, CancellationToken cancellationToken)
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
[18.884s] ConnectingFailedAsync: System.InvalidOperationException: Not allowed to connect while connect/disconnect is pending.
   at MQTTnet.Client.MqttClient.ConnectAsync(MqttClientOptions options, CancellationToken cancellationToken)
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
...
```
Version `4.3.3.952` after the fix:
```
[0.001s] Starting.
[0.231s] DisconnectedAsync.
[0.246s] ConnectingFailedAsync: MQTTnet.Exceptions.MqttCommunicationException: Client connected but server denied connection with reason 'NotAuthorized'.
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
[5.296s] DisconnectedAsync.
[5.296s] ConnectingFailedAsync: MQTTnet.Exceptions.MqttCommunicationException: Client connected but server denied connection with reason 'NotAuthorized'.
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
[10.333s] DisconnectedAsync.
[10.334s] ConnectingFailedAsync: MQTTnet.Exceptions.MqttCommunicationException: Client connected but server denied connection with reason 'NotAuthorized'.
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
[15.373s] DisconnectedAsync.
[15.373s] ConnectingFailedAsync: MQTTnet.Exceptions.MqttCommunicationException: Client connected but server denied connection with reason 'NotAuthorized'.
   at MQTTnet.Extensions.ManagedClient.ManagedMqttClient.ReconnectIfRequiredAsync(CancellationToken cancellationToken)
...
```

Seems like the fix works. You can try to reproduce and create new ticket.
